File: algorithms/1290/main.scala
import scala.math.pow

object Solution {
    def getDecimalValue(head: ListNode): Int = {
        val (ret, _) = recursive(head, 0)
        ret
    }
    
    def recursive(head: ListNode, count: Int): (Int, Int) = {
        (head.next match {
            case null => (0, count)
            case _ => recursive(head.next, count+1)
        }) match {
            case (x: Int, c: Int) => (pow(2, c-count).toInt * head.x + x, c)
        }
    }
}

File: algorithms/1290/main.go
import "math"

func getDecimalValue(head *ListNode) int {
    ret, _ := recursive(head, 0)
    return ret
}

func recursive(head *ListNode, count int) (int, int) {
    switch head.Next {
    case nil:
        return pow(2, 0) * head.Val, count
    default:
        ret, c := recursive(head.Next, count+1)
        return pow(2, c-count) * head.Val + ret, c
    }
}

func pow(num int, p int) int {
    return int(math.Pow(float64(num), float64(p)))
}
